### ISSUE TITLE
About reading and writing .op2 file

### ISSUE THREAD

**Issue by aa951742** [SteveDoyle2/pyNastran#621]
Hello Mr.Doyle,

I've been trying to modify my op2 file and output a new  op2 file to be read by Patran, and failing.
So I tried just reading the op2 in and immediately write it to a new op2 file without modifying it, and the new op2 file lacks data blocks GEOM1, GEOM2, GEOM3, GEOM4, EPT, MPT, and CASECC.
Data blocks OQG1, OUG1, OEF1X, OEF1, OES1X1, OES1, is included in the new op2 file.
After I imported the bdf file in Patran, and try to read my new op2 file, it successfully opened but failed to produce any result.
Did I read or write the op2 file the wrong way?
All i did is:
op2file = read_op2(op2_filename)
op2file.write_op2(r'D:\newop2.op2')

Sincerely,
Carl

**Comment by SteveDoyle2** [comment]
Carl,

Geometry is not included because you did not request it to be read.  You
need to use read_op2_geom if you also want the geometry to be written.

Do you have an example as well as an error message?

  I also don’t have Patran beyond 2005, so it’s a bit difficult to test.
The op2 is written in NX format, so that might be part of the issue.

Steve

**Comment by aa951742** [comment]
Thanks you for your answer Mr.Doyle,

So to include the geom data I need to use pyNastran.op2.op2_geom.read_op2_geom(op2_filename), correct?
How about CASECC, EPT, MPT?
After I use read_op2_geom to read my op2 file, and I output it to a new op2 file, it did include the GEOM1,2,3,4 data blocks, but when Patran reads the op2, it still failed.
The Patran2019 command window message is as follows:
$# The specified OUTPUT2 file "D:\project\try\4.op2" has been successfully opened.
$# Writing the OUTPUT2 information to the database, D:\Windows\Temp\05.db.
$# ***  Summary of OUTPUT2 File Translation
$# Total execution time in real seconds: 0
$# Total execution time in CPU seconds: 0
$# ***  End of Translation Summary
$# End of NASTRAN OUTPUT2 file translation.
The problem might be caused by the format had been changed since 2005 like you said.

Thank you again for your time,
Carl

**Comment by SteveDoyle2** [comment]
EPT and MPTare written of cards are found.  CASECC is not written.

Not writing the geometry tables should be equivalent to using
PARAM,OGEOM,NO.  The OP2 geometry Isn’t perfect (e.g., the CBAR/CBEAM won’t
be the same, some cards may not be written depending on the Nastran
version), but depending what you need it for, that might be ok.

Some tables (e.g., CASECC) are not supported.  Again, that shouldn’t affect
Patran’s ability to read the deck.

That’s odd that Patran read it successfully.  Could you run...

   test_op2 -c 4.op2

That will list what result types you have.  Also, can you share your
example?

**Comment by SteveDoyle2** [comment]
To be clear, in general removing tables out of the deck should make it work better.

I did find a bug in the MSC header (they include what I think is a date), as well an issue with the LAMA (eigenvalues) and grid point forces results (~OGFB1).  I fixed the MSC header issue, but haven't gotten to the others.  I haven't committed it quite yet.

I expect there will also be some issues with some OES/OEF results, but so far so good...
